src/mongocrypt.h.in

```diff
@@ -873,8 +888,9 @@ typedef enum {
    MONGOCRYPT_CTX_NEED_MONGO_MARKINGS = 2, /* run on mongocryptd. */
    MONGOCRYPT_CTX_NEED_MONGO_KEYS = 3,     /* run on key vault */
    MONGOCRYPT_CTX_NEED_KMS = 4,
+   MONGOCRYPT_CTX_NEED_KMS_CREDENTIALS = 7, /* fetch/renew KMS credentials */
```

Comment:
> @kevinAlbs Just thinking in terms of code complexity here, I think it would be simplest if `mongocrypt_setopt_kms_providers()` were to store a copy of the raw input BSON buffer, and then that could be parsed during each `mongocrypt_ctx_provide_kms_providers()` call before parsing that call’s own input BSON buffer, effectively merging the two KMS provider structs together? That seems a bit simpler than custom cloning/merging code in C

If storing the raw input BSON buffer simplifies merging that SGTM. Alternatively, a new function to merge could be defined `_mongocrypt_opts_kms_providers_merge`. `_mongocrypt_kek_copy_to` does something akin to that for KEK documents. Either approach seems OK.

Reply:
> What does the state machine do if `mongocrypt_setopt_use_need_kms_credentials_state` is invoked, but the `kmsProviders` map already contains credentials for the provider? My inclination is to say that in that case the `MONGOCRYPT_CTX_NEED_KMS_CREDENTIALS` state is not entered. This matches current behavior for `MONGODB-AWS` authentication.

If `mongocrypt_setopt_use_need_kms_credentials_state` is invoked, my original idea was to always request KMS credentials, regardless of whether they were set on `mongocrypt_t`.

However, I cannot think of a compelling use case for setting the "aws" KMS provider on `mongocrypt_t`, and then overriding it later.

If it does not add much complexity, and is consistent with the `MONGODB-AWS` callback, I am in favor of changing the implementation to only request credentials for a KMS provider if they are not set on `mongocrypt_t`